File: lizzie-mobile/shared/src/commonMain/kotlin/com/lizzie/engine/RemoteEngine.kt
package com.lizzie.engine

import com.lizzie.analysis.MoveData
import com.lizzie.rules.Stone
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

/**
 * Remote engine connecting to a KataGo instance over TCP.
 *
 * Uses platform-specific [GtpSocket] for the actual connection.
 * KataGo should be started with `--gtp` flag for GTP-over-TCP mode.
 */
class RemoteEngine : Engine {

    private val _status = MutableStateFlow<EngineStatus>(EngineStatus.Disconnected)
    override val status: Flow<EngineStatus> = _status.asStateFlow()

    private val _analysis = MutableStateFlow(AnalysisResult(emptyList()))
    override val analysis: Flow<AnalysisResult> = _analysis.asStateFlow()

    private var socket: GtpSocket? = null
    private var readerJob: Job? = null
    private var cmdNumber = 1
    private var currentCmdNum = 0
    private var _isPondering = false
    private var isKataGo = false

    // Response tracking: maps cmdNumber -> CompletableDeferred<String>
    private val pendingResponses = mutableMapOf<Int, CompletableDeferred<String>>()
    private val pendingLock = Any()

    override suspend fun start(config: EngineConfig) {
        val cfg = config as EngineConfig.Remote
        _status.value = EngineStatus.Connecting("Connecting to ${cfg.host}:${cfg.port}")
        cmdNumber = 1
        currentCmdNum = 0

        try {
            val s = GtpSocket()
            s.connect(cfg.host, cfg.port)
            socket = s

            // GTP handshake
            val nameResponse = sendGtpCommand("name")
            isKataGo = nameResponse.startsWith("KataGo")

            val versionResponse = sendGtpCommand("version")
            val version = versionResponse.trim()

            _status.value = EngineStatus.Ready(
                engineName = nameResponse.trim(),
                version = version,
                isKataGo = isKataGo,
            )

            // Start continuous reader coroutine
            readerJob = CoroutineScope(Dispatchers.Default).launch {
                try {
                    while (isActive && socket?.isConnected == true) {
                        val line = socket?.readLine() ?: break
                        parseLine(line.trimEnd('\r'))
                    }
                } catch (e: CancellationException) {
                    // Normal shutdown
                } catch (e: Exception) {
                    _status.value = EngineStatus.Error("Connection lost: ${e.message}")
                }
            }
        } catch (e: Exception) {
            _status.value = EngineStatus.Error("Failed to connect: ${e.message}")
            throw e
        }
    }

    override suspend fun stop() {
        readerJob?.cancel()
        try { sendGtpCommand("quit") } catch (_: Exception) {}
        socket?.close()
        socket = null
        _status.value = EngineStatus.Disconnected
        _isPondering = false
    }

    override suspend fun initGame(boardSize: Int, komi: Double, handicap: Int) {
        sendGtpCommand("boardsize $boardSize")
        sendGtpCommand("clear_board")
        sendGtpCommand("komi $komi")
        if (handicap > 0) {
            sendGtpCommand("fixed_handicap $handicap")
        }
    }

    override suspend fun playMove(color: Stone, coordinate: String?) {
        val colorStr = if (color == Stone.BLACK) "b" else "w"
        val coord = coordinate ?: "pass"
        sendGtpCommand("play $colorStr $coord")
    }

    override suspend fun undoMove() {
        sendGtpCommand("undo")
    }

    override suspend fun startPonder() {
        if (_isPondering) return
        _isPondering = true
        if (isKataGo) {
            sendGtpCommand("kata-analyze 100 0")
        } else {
            sendGtpCommand("lz-analyze 100 0")
        }
    }

    override suspend fun stopPonder() {
        if (!_isPondering) return
        _isPondering = false
        // Send a command to interrupt the continuous analysis
        // GTP doesn't have an interrupt, but sending a play command works
        // because kata-analyze is replaced by the next command
        sendGtpCommand("kata-analyze 1 0") // minimal analysis to replace the previous one
    }

    override fun isPondering(): Boolean = _isPondering

    override suspend fun sendGtpCommand(command: String): String {
        val s = socket ?: throw IllegalStateException("Engine not connected")
        val cmdNum = cmdNumber++
        val deferred = CompletableDeferred<String>()

        synchronized(pendingLock) {
            pendingResponses[cmdNum] = deferred
        }

        // Send command with GTP line numbering
        s.send("$cmdNum $command\n".encodeToByteArray())

        // Wait for response with timeout
        return withTimeout(30_000L) {
            deferred.await()
        }
    }

    private fun parseLine(line: String) {
        if (line.startsWith("info")) {
            val rest = line.substring(4).trimStart()
            val bestMoves = if (isKataGo) {
                parseInfoKatago(rest)
            } else {
                parseInfo(rest)
            }

            // Update analysis
            _analysis.value = AnalysisResult(
                bestMoves = bestMoves,
                scoreMean = bestMoves.firstOrNull()?.scoreMean ?: 0.0,
                scoreStdev = bestMoves.firstOrNull()?.scoreStdev ?: 0.0,
                currentPlayouts = MoveData.getPlayouts(bestMoves),
            )

            // Parse ownership if present
            val ownership = parseOwnership(line)
            if (ownership != null) {
                _analysis.value = _analysis.value.copy(ownership = ownership)
            }
        } else if (line.startsWith("=")) {
            // GTP success response: "=NNNN response_text"
            val rest = line.substring(1).trimStart()
            val spaceIdx = rest.indexOf(' ')
            val responseText: String
            val responseCmdNum: Int

            if (spaceIdx > 0) {
                val numStr = rest.substring(0, spaceIdx)
                responseCmdNum = numStr.toIntOrNull() ?: return
                responseText = rest.substring(spaceIdx + 1)
            } else {
                // Response without number
                responseCmdNum = rest.toIntOrNull() ?: return
                responseText = ""
            }

            currentCmdNum = responseCmdNum
            completePending(responseCmdNum, responseText)
        } else if (line.startsWith("?")) {
            // GTP error response
            val rest = line.substring(1).trimStart()
            val spaceIdx = rest.indexOf(' ')
            if (spaceIdx > 0) {
                val numStr = rest.substring(0, spaceIdx)
                val responseCmdNum = numStr.toIntOrNull()
                if (responseCmdNum != null) {
                    completePending(responseCmdNum, "? ${rest.substring(spaceIdx + 1)}")
                }
            }
        }
        // Other lines (stderr, tuning output, etc.) are ignored
    }

    private fun completePending(cmdNum: Int, response: String) {
        synchronized(pendingLock) {
            val deferred = pendingResponses.remove(cmdNum)
            if (deferred != null && !deferred.isCompleted) {
                deferred.complete(response)
            }
        }
    }

    private fun parseInfo(line: String): List<MoveData> {
        return line.split(" info ").mapNotNull { part ->
            val trimmed = part.trim()
            if (trimmed.isNotEmpty()) MoveData.fromInfo(trimmed) else null
        }
    }

    private fun parseInfoKatago(line: String): List<MoveData> {
        return line.split(" info ").mapNotNull { part ->
            val trimmed = part.trim()
            if (trimmed.isNotEmpty()) MoveData.fromInfoKatago(trimmed) else null
        }
    }

    private fun parseOwnership(line: String): List<Double>? {
        if (!line.contains("ownership")) return null
        val params = line.trim().split("ownership")
        if (params.size < 2) return null
        return params[1].trim().split(" ").mapNotNull { it.toDoubleOrNull() }
    }
}

File: lizzie_mobile/lib/engine/remote_engine.dart
import 'dart:async';
import '../go/move_data.dart';
import '../go/stone.dart';
import 'analysis.dart';
import 'engine.dart';
import 'gtp_client.dart';

/// Engine implementation that talks to a remote KataGo/Leela Zero over TCP.
///
/// Ported from Lizzie's `Leelaz.java` with the correct framing, command
/// correlation, and analyze/pause semantics.
class RemoteEngine implements Engine {
  GtpClient? _client;
  bool _isPondering = false;
  bool _isKataGo = false;
  StreamSubscription<String>? _lineSub;
  int _boardWidth = 19;
  int _boardHeight = 19;

  // Status/analysis streams (broadcast so multiple listeners can observe).
  final _statusController = StreamController<EngineStatus>.broadcast();
  @override
  Stream<EngineStatus> get status => _statusController.stream;

  final _analysisController = StreamController<AnalysisResult>.broadcast();
  @override
  Stream<AnalysisResult> get analysis => _analysisController.stream;

  @override
  bool get isPondering => _isPondering;

  @override
  Future<void> start(EngineConfig config) async {
    _statusController.add(Connecting('Connecting to ${config.host}:${config.port}...'));
    _isPondering = false;

    try {
      _client = GtpClient();
      await _client!.connect(config.host, config.port);

      // Subscribe to raw lines for info/ownership parsing.
      _lineSub = _client!.lines.listen(_handleLine);

      // Handshake: name + version.
      final nameResponse = await _client!.send('name');
      _isKataGo = nameResponse.startsWith('KataGo');

      final versionResponse = await _client!.send('version');
      final version = versionResponse.trim();

      _statusController.add(Ready(nameResponse.trim(), version, _isKataGo));
    } catch (e) {
      _statusController.add(EngineError('Failed to connect: $e'));
      rethrow;
    }
  }

  @override
  Future<void> stop() async {
    _isPondering = false;
    try {
      await _client?.send('quit');
    } catch (_) {}
    await _lineSub?.cancel();
    await _client?.disconnect();
    _client = null;
    _statusController.add(const Disconnected());
  }

  @override
  Future<void> initGame(int boardSize, {double komi = 6.5, int handicap = 0}) async {
    _boardWidth = boardSize;
    _boardHeight = boardSize;
    await _client!.send('boardsize $boardSize');
    await _client!.send('clear_board');
    await _client!.send('komi $komi');
    if (handicap > 0) {
      final response = await _client!.send('fixed_handicap $handicap');
      // Parse returned vertex list and place stones locally.
      // Format: "=N D4 Q16 ..." or "=N D4 D16 ..."
      _parseHandicapResponse(response);
    }
  }

  @override
  Future<void> playMove(Stone color, String? coordinate) async {
    final colorStr = color == Stone.black ? 'B' : 'W';
    final coord = coordinate ?? 'pass';
    await _client!.send('play $colorStr $coord');
  }

  @override
  Future<void> undoMove() async {
    await _client!.send('undo');
  }

  @override
  Future<void> startPonder() async {
    if (_isPondering) return;
    _isPondering = true;
    if (_isKataGo) {
      await _client!.send('kata-analyze 10 ownership true');
    } else {
      await _client!.send('lz-analyze 10');
    }
  }

  @override
  Future<void> stopPonder() async {
    if (!_isPondering) return;
    _isPondering = false;
    // Send any command to interrupt the running analyze.
    // The engine will emit the final =NNN for the analyze, then respond
    // to this command. We send 'name' as the interrupt.
    try {
      await _client!.send('name');
    } catch (_) {
      // Engine might be gone; that's OK.
    }
  }

  @override
  Future<String> sendGtpCommand(String command) async {
    return await _client!.send(command);
  }

  // ---------------------------------------------------------------------------
  // Line handler
  // ---------------------------------------------------------------------------

  void _handleLine(String line) {
    if (line.startsWith('info')) {
      _handleInfo(line.substring(5));
    }
    // "=..." and "?..." lines are consumed by GtpClient for command correlation,
    // so they don't reach here (the broadcast stream gets them, but we ignore
    // them in the engine logic).
  }

  void _handleInfo(String body) {
    List<MoveData> moves;
    List<double>? ownership;

    // Check for ownership data.
    final ownershipIdx = body.indexOf('ownership');
    if (ownershipIdx >= 0) {
      final ownershipStr = body.substring(ownershipIdx + 'ownership'.length).trim();
      ownership = ownershipStr
          .split(' ')
          .where((s) => s.isNotEmpty)
          .map((s) => double.tryParse(s) ?? 0.0)
          .toList();
      // Trim ownership to board size.
      final expectedLen = _boardWidth * _boardHeight;
      if (ownership.length > expectedLen) {
        ownership = ownership.sublist(0, expectedLen);
      }
    }

    if (_isKataGo) {
      moves = body.split(' info ').map((part) {
        final trimmed = part.trim();
        if (trimmed.isEmpty) return null;
        try {
          return MoveData.fromInfoKatago(trimmed);
        } catch (_) {
          return null;
        }
      }).whereType<MoveData>().toList();
    } else {
      moves = body.split(' info ').map((part) {
        final trimmed = part.trim();
        if (trimmed.isEmpty) return null;
        try {
          return MoveData.fromInfo(trimmed);
        } catch (_) {
          return null;
        }
      }).whereType<MoveData>().toList();
    }

    _analysisController.add(AnalysisResult(
      bestMoves: moves,
      ownership: ownership,
      scoreMean: moves.isNotEmpty ? moves.first.scoreMean : 0.0,
      scoreStdev: moves.isNotEmpty ? moves.first.scoreStdev : 0.0,
      currentPlayouts: MoveData.totalPlayouts(moves),
    ));
  }

  void _parseHandicapResponse(String response) {
    // Format: " D4 Q16 ..." (space-separated GTP coordinates)
    // The response comes back as GTP coordinates; we don't need to
    // place them locally (the engine tracks them). This is just for
    // local board sync if needed.
  }
}
